// chivox_h5sdk/src/controlpanel.js


import Utils from './utils';
import Config from './config.json';

/**
 * ControlPanel类。<br/><br/>
 *
 * ChiVoxSDK初始化时，录音机和播放器所需的控制面板。<br/>
 * 有播放，停止播放，录音，停止录音，回放录音，停止回放等功能。(仅用于控制UI控件状态)
 */
class ControlPanel {
    constructor() {
        this.ele_records = Utils.find_all(document, Config.selector.panel.record) || [];
        this.ele_plays = Utils.find_all(document, Config.selector.panel.play) || [];
        this.ele_replays = Utils.find_all(document, Config.selector.panel.replay) || [];
    }

    /**
     * 停止播放。
     * @param {Object=} curr_ele - 要停止播放的HTML DOM节点。当前页面有多个控制面板时需要指定。
     */
    playOff(curr_ele) {
        if (curr_ele) {
            Utils.removeClass(curr_ele, 'playOn');
            Utils.addClass(curr_ele, 'playOff');
        } else {
            for (var i = 0; i < this.ele_plays.length; i++) {
                Utils.removeClass(this.ele_plays[i], 'playOn');
                Utils.addClass(this.ele_plays[i], 'playOff');
            }
        }
    }

    /**
     * 开始播放。
     * @param {Object=} curr_ele - 要开始播放的HTML DOM节点。当前页面有多个控制面板时需要指定。
     */
    playOn(curr_ele) {
        if (!curr_ele && this.ele_plays.length > 0) {
            curr_ele = this.ele_plays[0];
        }

        if (curr_ele) {
            Utils.removeClass(curr_ele, 'playOff');
            Utils.addClass(curr_ele, 'playOn');
        }
    }

    /**
     * 停止录音。
     * @param {Object=} curr_ele - 要停止录音的HTML DOM节点。当前页面有多个控制面板时需要指定。
     */
    recordOff(curr_ele) {
        if (curr_ele) {
            Utils.removeClass(curr_ele, 'recordOn');
            Utils.addClass(curr_ele, 'recordOff');
        } else {
            for (var i = 0; i < this.ele_records.length; i++) {
                Utils.removeClass(this.ele_records[i], 'recordOn');
                Utils.addClass(this.ele_records[i], 'recordOff');
            }
        }
    }

    /**
     * 开始录音。
     * @param {Object=} curr_ele - 要开始录音的HTML DOM节点。当前页面有多个控制面板时需要指定。
     */
    recordOn(curr_ele) {
        if (!curr_ele && this.ele_records.length > 0) {
            curr_ele = this.ele_records[0];
        }

        if (curr_ele) {
            Utils.removeClass(curr_ele, 'recordOff');
            Utils.addClass(curr_ele, 'recordOn');
        }
    }

    /**
     * 禁用回放功能。
     * @param {Object=} curr_ele - 要禁止回放的HTML DOM节点。当前页面有多个控制面板时需要指定。
     */
    replayDisabled(curr_ele) {
        if (curr_ele) {
            Utils.removeClass(curr_ele, 'replayOn');
            Utils.removeClass(curr_ele, 'replayOff');
            Utils.addClass(curr_ele, 'replayDisabled');

            curr_ele.setAttribute('dataRecordId', '');
        } else {
            for (var i = 0; i < this.ele_replays.length; i++) {
                curr_ele = this.ele_replays[i];
                Utils.removeClass(curr_ele, 'replayOn');
                Utils.removeClass(curr_ele, 'replayOff');
                Utils.addClass(curr_ele, 'replayDisabled');

                curr_ele.setAttribute('dataRecordId', '');
            }
        }
    }

    /**
     * 停止回放功能。
     * @param {Object=} curr_ele - 要停止回放的HTML DOM节点。当前页面有多个控制面板时需要指定。
     */
    replayOff(curr_ele) {
        if (curr_ele) {
            Utils.removeClass(curr_ele, 'replayOn');
            Utils.removeClass(curr_ele, 'replayDisabled');
            Utils.addClass(curr_ele, 'replayOff');
        } else {
            for (var i = 0; i < this.ele_replays.length; i++) {
                curr_ele = this.ele_replays[i];
                Utils.removeClass(curr_ele, 'replayOn');
                Utils.removeClass(curr_ele, 'replayDisabled');
                Utils.addClass(curr_ele, 'replayOff');
            }
        }
    }

    /**
     * 开始回放功能。
     * @param {Object=} curr_ele - 要开始回放的HTML DOM节点。当前页面有多个控制面板时需要指定。
     */
    replayOn(curr_ele) {
        if (!curr_ele && this.ele_plays.length > 0) {
            curr_ele = this.ele_replays[0];
        }
        if (curr_ele) {
            Utils.removeClass(curr_ele, 'replayOff');
            Utils.removeClass(curr_ele, 'replayDisabled');
            Utils.addClass(curr_ele, 'replayOn');
        }
    }
}

export default ControlPanel;
